cstrml/payment/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+    
+    /// Used to transfer
+    type CurrencyToBalance: Convert<BalanceOf<Self>, u128> + Convert<u128, BalanceOf<Self>>;
+
+    type BalanceInterface: self::BalanceInterface<Self::Origin, Self::AccountId, BalanceOf<Self>>;
+}
+
+// This module's storage items.
+decl_storage! {
+    trait Store for Module<T: Trait> as Market {
+        /// A mapping from storage provider to order id
+        pub Payments get(fn payments):
+        map hasher(twox_64_concat) T::Hash => Option<PaymentLedger<BalanceOf<T>>>;
+    }
+}
+
+decl_error! {
+    /// Error for the market module.
+    pub enum Error for Module<T: Trait> {
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        // Initializing events
+        // this is needed only if you are using events in your module
+        fn deposit_event() = default;
+
+        /// Enact a proposal from a referendum. For now we just make the weight be the maximum.
+		#[weight = 1_000_000]
+        fn payment_by_instalments(
+            origin,
+            client: <T as system::Trait>::AccountId,
+            provider: <T as system::Trait>::AccountId,
+            value: BalanceOf<T>,
+            order_id: T::Hash
+        ) -> DispatchResult {
+            ensure_root(origin.clone())?;
+            // 0. check left currency
+            let payment_ledger = Self::payments(order_id).unwrap_or_default();
+            let real_value = value.min(payment_ledger.total - payment_ledger.already_paid);
+
+            if !Zero::is_zero(&real_value) {
+                // 1. unreserve one piece currency
+                T::Currency::unreserve(
+                    &client,
+                    real_value);
+                <Payments<T>>::mutate(&order_id, |payment_ledger| {
+                    if let Some(p) = payment_ledger {
+                        p.already_unreserved += real_value;
+                    }
+                });
+                // Check the order status
+                if let Some(sorder) = T::MarketInterface::maybe_get_sorder(&order_id) {
+                    match sorder.order_status {
+                        OrderStatus::Success => {
+                            // 3. (Maybe) transfer the currency
+                            if T::BalanceInterface::maybe_transfer(origin.clone(), &client, &provider, real_value) {
+                                // 4. update payments
+                                <Payments<T>>::mutate(&order_id, |payment_ledger| {
+                                    if let Some(p) = payment_ledger {
+                                        p.already_paid += real_value;
+                                    }
+                                });
+                                Self::deposit_event(RawEvent::PaymentSuccess(client));
+                            }
+
+                        },
+                        // TODO: Deal with failure status
```

Comment:
Delete this `TODO`, cause we won't deal with failure status in `Payment` module, then this `match` can simplify.